### PR TITLE
Making compute_loss_func always take priority in Trainer

### PR DESCRIPTION
**Pull request by Flakes342**
# What does this PR do?

Fixes #40601 

Currently, Trainer only uses compute_loss_func when labels are provided. This leads to unintuitive behavior:

1. User-defined losses are silently ignored if labels=None
2. Advanced training setups (unsupervised, contrastive, RLHF) cannot rely on custom losses consistently
3. Debugging is difficult since the function is skipped without notice

With this PR, now:
1. `compute_loss_func` always takes priority, regardless of whether labels are passed
2. If labels=None, the function is still called with labels=None
3. Adds a temporary warning to highlight the change 

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [x] Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/main/CONTRIBUTING.md#create-a-pull-request),
      Pull Request section?
- [x] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
      to it if that's the case.  #40601 
- [ ] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/transformers/tree/main/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/main/docs#writing-source-documentation).
- [ ] Did you write any new necessary tests?

## Who can review?

Anyone in the community is free to review the PR once the tests have passed. Feel free to tag
members/contributors who may be interested in your PR.

### CODE REVIEW

**Comment by Rocketknight1** [comment]
cc @sunmarc

**Comment by Flakes342** [comment]
@SunMarc the PR is ready now..awaiting your review. Thanks!

File: src/transformers/trainer.py [hunk 1]
```diff
@@ -4129,27 +4129,41 @@ def compute_loss(
...
-                raise ValueError(
-                    "The model did not return a loss from the inputs, only the following keys: "
-                    f"{','.join(outputs.keys())}. For reference, the inputs it received are {','.join(inputs.keys())}."
+        # User-defined compute_loss function
+        if self.compute_loss_func is not None:
+            if labels is None:
+                warnings.warn(
+                    "Trainer: `compute_loss_func` is defined but `labels=None`. "
+                    "Your custom loss function will still be called with labels=None. ",
+                    UserWarning,
```
Comment (SunMarc) [inline src/transformers/trainer.py]: use logger.warning instead !

File: src/transformers/trainer.py [hunk 2]
```diff
@@ -4129,27 +4129,41 @@ def compute_loss(
...
                 )
-            # We don't use .loss here since the model may return tuples instead of ModelOutput.
-            loss = outputs["loss"] if isinstance(outputs, dict) else outputs[0]
+            loss = self.compute_loss_func(
+                outputs,
+                labels,
+                num_items_in_batch=num_items_in_batch,
+            )
+        else:
+            if labels is not None:
```
Comment (SunMarc) [inline src/transformers/trainer.py]: change into a elif and unindent

**Review by SunMarc — COMMENTED** [review]
Thanks for the PR ! Just a few nits

**Comment by SunMarc** [comment]
Can you rebase correctly ? There are some unrelated changes that have already been upstreamed

**Comment by Flakes342** [comment]
yes, I saw that..have fixed it now, waiting for unit test results

**Comment by Flakes342** [comment]
Hi Marc, any idea why it is failing?

File: src/transformers/trainer.py [hunk 3]
```diff
@@ -4117,7 +4117,19 @@ def compute_loss(
...
+            if labels is None:
+                logger.warning(
+                    "Trainer: `compute_loss_func` is defined but `labels=None`. "
+                    "Your custom loss function will still be called with labels=None. "
+                )
+            loss = self.compute_loss_func(
+                outputs,
+                labels,
+                num_items_in_batch=num_items_in_batch,
+            )
```
Comment (SunMarc) [inline src/transformers/trainer.py:4129]: you forgot to remove the lines below since you added this

**Review by SunMarc — APPROVED** [review]
Nice !
